File: Go/Counter/logger.go
package main

import (
	"bufio"
	logger "counter/Plugin"
	"fmt"
	"net"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"

	"google.golang.org/protobuf/proto"
)

type LamportClock struct {
	lClock int
}

func NewLamportClock() *LamportClock {
	return &LamportClock{lClock: 0}
}

func (lc *LamportClock) Increment() {
	lc.lClock++
}

func (lc *LamportClock) GetTimestamp() (int, time.Time) {
	return lc.lClock, time.Now()
}

var lc *LamportClock

type ReplicaState struct {
	VersionMapState map[int]int
	LastVersion     int
}

var replicaStates = make(map[int]*ReplicaState)

func create_log(rId int) {
	logFile := fmt.Sprintf("Replica_%d.log", rId)

	if _, err := os.Stat(logFile); err == nil {
		fmt.Println("Log File Already Exists")
	}

	file, err := os.Create(logFile)
	if err != nil {
		fmt.Errorf("failed to create file: %w", err)
	}

	defer file.Close()

	lc = NewLamportClock()

	replicaStates[rId] = &ReplicaState{
		VersionMapState: make(map[int]int),
		LastVersion:     0,
	}
	// init version
	replicaStates[rId].VersionMapState[0] = 0
	replicaStates[rId].LastVersion++
}

func persist(rId int, msg string) {
	logFile := fmt.Sprintf("Replica_%d.log", rId)

	// append mode
	file, err := os.OpenFile(logFile, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		fmt.Errorf("failed to open file: %w", err)
		return
	}
	defer file.Close()

	lc.Increment()
	lamportTime, physicalTime := lc.GetTimestamp()

	logEntry := fmt.Sprintf("%s, Lamport Time: %d, Physical Time: %s\n",
		msg, lamportTime, physicalTime.Format(time.RFC3339))
	if _, err := file.WriteString(logEntry); err != nil {
		fmt.Errorf("failed to log: %w", err)
	}

	if strings.Contains(msg, "Updated Value is") {
		state, _ := replicaStates[rId]
		state.VersionMapState[state.LastVersion] = extractVersionValue(msg)
		state.LastVersion++
	}
}

func execute(bytes []byte) {

	var log logger.LogMsg
	err := proto.Unmarshal(bytes, &log)

	if err != nil {
		fmt.Errorf("error in unmarshalling, ", err)
	}

	id, logMsg := int(log.GetId()), log.GetLogs()

	if strings.Contains(logMsg, "Undo") {
		tasks := strings.Split(logMsg, "_")
		task := tasks[1]
		// fmt.Println("Undo taks is ", task)
		fmt.Println("Undo request came from application")
		action := undo(task)
		// fmt.Println("counter action is ", action)
		replica_addr := ""

		counter_action := &logger.LogMsg{
			Id:   int32(id),
			Logs: string(action),
		}
		serialized, err := proto.Marshal(counter_action)
		if err != nil {
			fmt.Errorf("error in marshalling, ", err)
		}

		serialized = append(serialized, 0x00)
		if id == 1 {
			replica_addr = "localhost:8081"
		} else if id == 2 {
			replica_addr = "localhost:8082"
		} else {
			replica_addr = "localhost:8083"
		}
		sendBacktoReplica(replica_addr, serialized)
	} else if strings.Contains(logMsg, "Rev") {
		tasks := strings.Split(logMsg, "_")
		rev := tasks[1]
		// fmt.Println("Undo taks is ", task)
		fmt.Println("Rollback request came from application: ", rev)
		action := reversibility(id, rev)
		// fmt.Println(replicaStates[id])
		fmt.Println("counter action is ", action)
		replica_addr := ""

		counter_action := &logger.LogMsg{
			Id:   int32(id),
			Logs: string(action),
		}
		serialized, err := proto.Marshal(counter_action)
		if err != nil {
			fmt.Errorf("error in marshalling, ", err)
		}

		serialized = append(serialized, 0x00)
		if id == 1 {
			replica_addr = "localhost:8081"
		} else if id == 2 {
			replica_addr = "localhost:8082"
		} else {
			replica_addr = "localhost:8083"
		}
		sendBacktoReplica(replica_addr, serialized)

	} else if logMsg == "create" {
		create_log(id)
	} else {
		persist(id, logMsg)
	}
}

func main() {
	listener, err := net.Listen("tcp", "localhost:8080")
	if err != nil {
		fmt.Printf("Failed to listen on port 8080: %v\n", err)
		os.Exit(1)
	}
	defer listener.Close()

	fmt.Println("Logger server started on localhost:8080")

	for {
		conn, err := listener.Accept()
		if err != nil {
			fmt.Printf("Failed to accept connection: %v\n", err)
			continue
		}

		go keepConnection(conn)
	}
}

func keepConnection(conn net.Conn) {
	reader := bufio.NewReader(conn)
	// reqs, err := reader.ReadBytes('\n')
	var message []byte
	for {
		b, err := reader.ReadByte()
		if err != nil {
			// fmt.Println("Client left.")
			conn.Close()
			return
		}

		// Check for delimiter
		if b == 0x00 {
			break
		}

		message = append(message, b)
	}
	execute(message)
	keepConnection(conn)
}

func sendBacktoReplica(replica_addr string, message []byte) {
	conn, err := net.Dial("tcp", replica_addr)
	if err != nil {
		fmt.Printf("Failed to connect to logger: %v\n", err)
		return
	}
	defer conn.Close()

	_, err = conn.Write(message)
	if err != nil {
		fmt.Printf("Failed to send message: %v\n", err)
	}
}

func undo(task string) string {
	if task == "Inc" {
		return "Dec"
	}
	return "Inc"
}

func reversibility(id int, version string) string {
	state, _ := replicaStates[id]
	rev_version, _ := strconv.Atoi(version)
	rev_val := state.VersionMapState[rev_version]

	cur_val := state.VersionMapState[state.LastVersion-1]
	action := ""
	if cur_val == rev_val {
		fmt.Println("Rolledback version is the same")
	} else if cur_val > rev_val {
		diff := cur_val - rev_val
		action = fmt.Sprintf("Rev_%d_Dec", diff)
	} else {
		diff := rev_val - cur_val
		action = fmt.Sprintf("Rev_%d_Inc", diff)
	}
	return action
}

func extractVersionValue(s string) int {
	substring := "Updated Value is"

	// Regular expression to find digits following "Updated Value is"
	re := regexp.MustCompile(substring + ` (\d+)`)
	match := re.FindStringSubmatch(s)
	if len(match) > 1 {
		// Convert the matched number to an integer
		var value int
		fmt.Sscanf(match[1], "%d", &value)
		return value
	}

	// error value
	return 0
}

File: Go/Counter/logger.js
const { LogMsg } = require('./Plugin/logger_pb.js')
const net = require("net")
const fs = require("fs")
const path = require("path")


class LamportClock {
    constructor() {
        this.lClock = 0
    }

    increment() {
        this.lClock++
    }

    getTimestamp() {
        return { lamportTime: this.lClock, physicalTime: new Date() }
    }
}

let lc = null

class ReplicaState {
    constructor() {
        this.versionMapState = {}
        this.lastVersion = 0
        this.versionMapState[0] = 0
        this.lastVersion++
    }
}

let replicaStates = {}

function createLog(rId) {
    const logFile = `Replica_${rId}.log`

    if (fs.existsSync(logFile)) {
        console.log("Log File Already Exists")
    } else {
        fs.writeFileSync(logFile, "", { flag: "w" })
    }

    lc = new LamportClock()
    replicaStates[rId] = new ReplicaState()
}

function persist(rId, msg) {
    const logFile = `Replica_${rId}.log`

    lc.increment()
    const { lamportTime, physicalTime } = lc.getTimestamp()
    const logEntry = `${msg}, Lamport Time: ${lamportTime}, Physical Time: ${physicalTime.toISOString()}\n`

    fs.appendFile(logFile, logEntry, (err) => {
        if (err) console.error("Failed to log:", err)
    })

    if (String(msg).includes("Updated Value is")) {
        const state = replicaStates[rId]
        const value = extractVersionValue(msg)
        state.versionMapState[state.lastVersion] = value
        state.lastVersion++
    }
}

function execute(bytes) {
    const logMsg = LogMsg.deserializeBinary(bytes)

    const id = logMsg.getId()
    const logMessage = logMsg.getLogs()

    if (String(logMsg).includes("Undo")) {
        const tasks = String(logMsg).split("_")
        const task = tasks[1]

        console.log("Undo request came from application")
        const action = undo(task)

        let counterAction = new LogMsg()
        counterAction.setId(id)
        counterAction.setLogs(action)

        let serialized = counterAction.serializeBinary()
        serialized = Buffer.from([...serialized, 0x00])

        let replicaAddr
        if (id === 1) {
            replicaAddr = "localhost:8081"
        } else if (id === 2) {
            replicaAddr = "localhost:8082"
        } else {
            replicaAddr = "localhost:8083"
        }

        sendBackToReplica(replicaAddr, serialized)
    } else if (String(logMsg).includes("Rev")) {
        const tasks = String(logMsg).split("_")
        const rev = tasks[1]
        console.log("Rollback request came from application:", rev)

        const action = reversibility(id, rev)
        console.log("Counter action is:", action)

        let counterAction = new LogMsg()
        counterAction.setId(id)
        counterAction.setLogs(action)

        let serialized = counterAction.serializeBinary()
        serialized = Buffer.from([...serialized, 0x00])

        let replicaAddr
        if (id === 1) {
            replicaAddr = "localhost:8081"
        } else if (id === 2) {
            replicaAddr = "localhost:8082"
        } else {
            replicaAddr = "localhost:8083"
        }
        sendBackToReplica(replicaAddr, serialized)
    } else if (logMessage === "create") {
        createLog(id)
    } else {
        persist(id, logMessage)
    }
}

async function keepConnection(conn) {
    const reader = conn
    let message = Buffer.alloc(0)

    conn.on('data', (data) => {
        for (let i = 0; i < data.length; i++) {
            if (data[i] === 0x00) {
                execute(message)
                message = Buffer.alloc(0)
                return
            }
            message = Buffer.concat([message, Buffer.from([data[i]])])
        }
    })

    conn.on('end', () => {
        // console.log('Client left.')
        conn.end()
    })
}

async function main() {
    const server = net.createServer((conn) => {
        // console.log("Accepted new connection")
        keepConnection(conn)
    })

    server.listen(8080, "localhost", () => {
        console.log("Logger server started on localhost:8080")
    })

    server.on("error", (err) => {
        console.error("Server error:", err)
    })
}


function sendBackToReplica(replicaAddr, message) {
    const [host, port] = String(replicaAddr).split(":")

    const client = new net.Socket()
    client.connect(port, host, () => {
        client.write(message)
    })

    client.on('error', (err) => {
        console.error("Failed to connect to replica:", err.message)
    })

    client.on('close', () => {
        // console.log("Connection closed")
    })
}

function undo(task) {
    return task === "Inc" ? "Dec" : "Inc"
}

function reversibility(id, version) {
    const state = replicaStates[id]
    const revVersion = parseInt(version, 10)
    const revVal = state.versionMapState[revVersion]
    const curVal = state.versionMapState[state.lastVersion - 1]
    let action = ""

    if (curVal === revVal) {
        console.log("Rolled back version is the same")
    } else if (curVal > revVal) {
        const diff = curVal - revVal
        action = `Rev_${diff}_Dec`
    } else {
        const diff = revVal - curVal
        action = `Rev_${diff}_Inc`
    }

    return action
}

function extractVersionValue(s) {
    const match = String(s).match(/Updated Value is (\d+)/)
    return match ? parseInt(match[1], 10) : 0
}

main().catch(err => {
    console.error('Error:', err.message)
    process.exit(1)
})